### PR TITLE
Update k8s to 1.15 and us go modules

### LINKED ISSUE #56: Use go modules
We should use go modules to keep in lined with Kubernetes upstream

### PR DESCRIPTION
**Pull request by JoshVanL**
Signed-off-by: JoshVanL <vleeuwenjoshua@gmail.com>

/assign

fixes #56 

```release-note
Updates vendored Kubernetes to 1.15 and uses Go modules
```

### CODE REVIEW

**Comment by JoshVanL** [comment]
All green :sunglasses: 
/unassign
/assign @simonswine

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.0.0-20190620085942-b7f18460b210
+	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20190228160746-b3a7cee44a3
+	k8s.io/kube-proxy => k8s.io/kube-proxy v0.0.0-20190620085809-589f994ddf7f
+	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.0.0-20190620085912-4acac5405ec6
+	k8s.io/kubectl => k8s.io/kubectl v0.0.0-20190602132728-7075c07e78bf
+	k8s.io/kubelet => k8s.io/kubelet v0.0.0-20190620085838-f1cb295a73c9
+	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20190620090156-2138f2c9de18
+	k8s.io/metrics => k8s.io/metrics v0.0.0-20190620085625-3b22d835f165
+	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20190620085408-1aef9010884e
+	sigs.k8s.io/structured-merge-diff => sigs.k8s.io/structured-merge-diff v0.0.0-20190302045857-e85c7b244fd2
```
Comment (munnerz) [inline go.mod]: Looks like you are pinning a lot more than is needed here - you can remove all of these that are not listed in the `require` block above 😄

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+replace (
+	github.com/golang/mock => github.com/golang/mock v1.3.1
+	k8s.io/kubernetes => k8s.io/kubernetes v1.15.0
```
Comment (munnerz) [inline go.mod]: Do you depend on k8s.io/kubernetes? 🤔

File: pkg/e2e/e2e_test.go [hunk 3]
```diff
@@ -64,31 +66,33 @@ func TestMain(m *testing.M) {
...
+			kubeadmConfig = fmt.Sprint(`apiVersion: kubeadm.k8s.io/v1alpha3
 kind: ClusterConfiguration
 `, kubeadmConfig)
-	}
+		}
 
-	conf.KubeadmConfigPatches = []string{kubeadmConfig}
+		conf.KubeadmConfigPatches = []string{kubeadmConfig}
+	} else {
+		conf.Networking.ServiceSubnet = "10.0.0.0/16"
```
Comment (munnerz) [inline pkg/e2e/e2e_test.go]: Not sure if this is new code or not, but how come you need to set up a custom service CIDR?

**Comment by munnerz** [comment]
It may be worthwhile adding an [update-vendor.sh](https://github.com/jetstack/cert-manager/blob/master/hack/update-vendor.sh) and [verify-vendor.sh](https://github.com/jetstack/cert-manager/blob/master/hack/verify-vendor.sh) to ensure that the `vendor/`, `go.mod` and `go.sum` files are up to date.

One key difference is that the `update-vendor.sh` script above runs `go mod tidy`, which tidies up your go.mod file as well as resolves any unresolved import references (i.e. converts branch names to their full deterministic form)

File: pkg/e2e/e2e_test.go (hunk 3 above)
Comment (JoshVanL) [reply pkg/e2e/e2e_test.go]: Was causing circular DNS issues in CI

File: go.mod (hunk 2 above)
Comment (JoshVanL) [reply go.mod]: Is there an issue with that..?

**Comment by JoshVanL** [comment]
I have now removed `k8s.io/kubernetes` which is great.
I have had to manually copy over OIDC auth flags which is...fine.

I also had to make some changes to the upgrade tests to just reply on client-go and co. which is a bit better now.

Assuming tests pass can you take a look?
/cc @simonswine @munnerz 
/test pull-kube-oidc-proxy-e2e-v1-14
/test pull-kube-oidc-proxy-e2e-v1-13
/test pull-kube-oidc-proxy-e2e-v1-12
/test pull-kube-oidc-proxy-e2e-v1-11

**Comment by JoshVanL** [comment]
for {
/retest
}

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	github.com/spf13/pflag v1.0.3
+	gopkg.in/square/go-jose.v2 v2.3.1
+	k8s.io/api v0.0.0
+	k8s.io/apimachinery v0.0.0
+	k8s.io/apiserver v0.0.0-20190721103406-1e59c150c171
+	k8s.io/cli-runtime v0.0.0
+	k8s.io/client-go v11.0.0+incompatible
+	k8s.io/component-base v0.0.0
+	k8s.io/klog v0.3.3
+	sigs.k8s.io/kind v0.0.0-00010101000000-000000000000
```
Comment (munnerz) [inline go.mod]: I've seen this happen before with kind and I'm not 100% certain why.. I've you run `./hack/pin-dependency.sh sigs.k8s.io/kind v0.4.0`, I think it should update this string to be correct too. I don't think it's that important to do though given the `replace` directive + go.sum file.

File: pkg/e2e/upgrade_test.go [hunk 5]
```diff
@@ -119,97 +118,108 @@ func Test_Upgrade(t *testing.T) {
...
+			Command: []string{
+				"curl", "127.0.0.1:8080", "-s", "-d", "hello world",
+			},
+			Stdin:  false,
+			Stdout: true,
+			Stderr: true,
+			TTY:    false,
+		}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(restConfig, "POST", req.URL())
```
Comment (munnerz) [inline pkg/e2e/upgrade_test.go]: Out of interest, does this mean a change in the supported k8s versions? What versions support SPDY?

Comment (JoshVanL) [reply pkg/e2e/upgrade_test.go]: Only the k8s version we support :man_shrugging: 
1.11->15

File: go.mod (hunk 4 above)
Comment (JoshVanL) [reply go.mod]: Changed it to 0.4.0 now. Just a cosmetic change in mod, no vendor changes

**Pull request merged**